[ cubical, doc, re #3539 ] Document the cubical mode properly (#3571)
agda/agda: doc/user-manual/language/cubical.lagda.rst

Setω`` (the reason this is in ``Setω`` is because it doesn't support
the ``transp`` and ``hcomp`` operations). A variable ``i : I``
intuitively corresponds to a point the `real unit interval
<https://en.wikipedia.org/wiki/Unit_interval>`_. In an empty context,
there are only two values of type ``I``: the two endpoints of the
interval, ``i0`` and ``i1``.

.. code-block:: agda

  i0 : I
  i1 : I

Elements of the interval form a `De Morgan algebra
<https://en.wikipedia.org/wiki/De_Morgan_algebra>`_, with minimum
(``∧``), maximum (``∨``) and negation (``~``).

.. code-block:: agda

  _∧_ : I → I → I
  _∨_ : I → I → I
  ~_ : I → I

All the properties of De Morgan algebras hold definitionally. The
endpoints of the interval ``i0`` and ``i1`` are the bottom and top
elements, respectively.

.. code-block:: agda

    i0 ∨ i    = i
    i  ∨ i1   = i1
    i  ∨ j    = j ∨ i
    i0 ∧ i    = i0
    i1 ∧ i    = i
    i  ∧ j    = j ∧ i
    ~ (~ i)   = i
    i0        = ~ i1
    ~ (i ∨ j) = ~ i ∧ ~ j
    ~ (i ∧ j) = ~ i ∨ ~ j

The core idea of Homotopy Type Theory and Univalent Foundations is a
correspondence between paths (as in topology) and (proof-relevant)

> equality (as in Martin-Löf's identity type). This correspondence is

Reviewer: ```suggestion
equalities (as in Martin-Löf's identity type). This correspondence is
```

  JRefl : ∀ {ℓ} {A : Set ℓ} {x : A} (P : ∀ y → x ≡ y → Set ℓ)
           (d : P x refl) → J P d refl ≡ d
  JRefl P d = transportRefl d

Internally in Agda the ``transp`` operation computes by cases on the
type, so for example for Σ-types it is computed elementwise. For path
types it is however not yet possible to provide the computation rule
as we need some way to remember the endpoints of the path after
transporting it. Furthermore, this must work for arbitrary higher
dimensional cubes (as we can iterate the path types). For this we
introduce the "homogeneous composition operations" (``hcomp``) that
generalize binary composition of paths to n-ary composition of higher
dimensional cubes.

Partial elements
----------------

In order to describe the homogeneous composition operations we need to
be able to write partially specified n-dimensional cubes (i.e. cubes
where some faces are missing). Given an element of the interval ``r :
I`` there is a predicate ``IsOne`` which represents the constraint ``r
= i1``. This comes with a proof that ``i1`` is in fact equal to ``i1``
called ``1=1 : IsOne i1``. We use Greek letters like ``φ`` or ``ψ``

> when such an ``r`` should be thought of as being in the image of

Reviewer: ```suggestion
when such an ``r`` should be thought of as being in the domain of
```
I think that’s what you meant?

Author: What I want to say is that when we write φ the argument should be thought as of being in the face lattice (which is not free as `(i=0) /\ (i=1) = \bot`). The IsOne map goes from the interval to the face lattice.

Reviewer: Ok, but from the point of view of Agda, `IsOne` goes from `I` to `Setω`, so it’s strange to say that `r` is in the image of `IsOne`, as `r` is of type `I`.

Author: Ah I see what you're saying. I'm always thinking in terms of the semantics. With this in mind I think "domain" is less confusing

``IsOne``.

Using this we introduce a type of partial elements called ``Partial φ
A``, this is a special version of ``IsOne φ → A`` with a more
extensional judgmental equality (two elements of ``Partial φ A`` are
considered equal if they represent the same subcube, so the faces of
the cubes can for example be given in different order and the two
elements will still be considered the same). The idea is that
``Partial φ A`` is the type of cubes in ``A`` that are only defined
when ``IsOne φ``.  There is also a dependent version of this called
``PartialP φ A`` which allows ``A`` to be defined only when ``IsOne
φ``.

.. code-block:: agda

  Partial : ∀ {ℓ} → I → Set ℓ → Setω

  PartialP : ∀ {ℓ} → (φ : I) → Partial φ (Set ℓ) → Setω

> Partial elements are introduced using pattern matching:

Reviewer: ```suggestion
There is a new form of pattern matching that can be used to introduce partial elements:
```
Clarify the sentence, following @WolframKahl’s comments.

::

  partialBool : ∀ i → Partial (i ∨ ~ i) Bool
  partialBool i (i = i0) = true
  partialBool i (i = i1) = false

The term ``partialBool`` should be thought of a boolean with different
values when ``(i = i0)`` and ``(i = i1)``. Terms of type ``Partial φ
A`` can also be introduced using pattern matching lambdas

> (http://wiki.portal.chalmers.se/agda/pmwiki.php?n=ReferenceManual.PatternMatchingLambdas).

Reviewer: The link should point to https://agda.readthedocs.io/en/latest/language/lambda-abstraction.html#pattern-matching-lambda.

Author: Fixed

::

  partialBool' : ∀ i → Partial (i ∨ ~ i) Bool
  partialBool' i = λ { (i = i0) → true
                     ; (i = i1) → false }

When the cases overlap they must agree (note that the order of the
cases doesn't have to match the interval formula exactly):

::

  partialBool'' : ∀ i j → Partial (~ i ∨ i ∨ (i ∧ j)) Bool
  partialBool'' i j = λ { (i = i1)          → true
                        ; (i = i1) (j = i1) → true
                        ; (i = i0)          → false }

Furthermore ``IsOne i0`` is actually absurd

::

  empty : {A : Set} → Partial i0 A
  empty = λ { () }

Cubical Agda also has cubical subtypes as in the CCHM type theory:

::

  _[_↦_] : ∀ {ℓ} (A : Set ℓ) (φ : I) (u : Partial φ A) → Setω
  A [ φ ↦ u ] = Sub A φ u

A term ``v : A [ φ ↦ u ]`` is of type ``A`` and when ``IsOne φ`` it

> must be definitionally equal to ``u : A``. Any term ``u : A`` can be

Reviewer: It’s not true that `v` is of type `A` (you need to use `ouc` for that), so I would rather say:

A term `v : A [ φ ↦ u ]` should be thought of as a term of type `A` which is definitionally equal to `u:A` when `IsOne φ` is satisfied.

Author: Fixed

.. code-block:: agda

  hcomp : ∀ {ℓ} {A : Set ℓ} {φ : I} (u : I → Partial φ A) (u0 : A) → A

When calling ``hcomp {φ = φ} u u0`` Agda makes sure that ``u0`` agrees
with ``u i0`` on ``φ``. The idea is that ``u0`` is the base and ``u``
specifies the sides of an open box. This is hence an open (higher
dimensional) cube where the side opposite of ``u0`` is missing. The
``hcomp`` operation then gives us the missing side opposite of
``u0``. For example binary composition of paths can be written as:

::

  compPath : ∀ {ℓ} {A : Set ℓ} {x y z : A} → x ≡ y → y ≡ z → x ≡ z
  compPath {x = x} p q i = hcomp (λ j → λ { (i = i0) → x
                                          ; (i = i1) → q j })
                                 (p i)

Pictorially we are given ``p : x ≡ y`` and ``q : y ≡ z``, and the
composite of the two paths is obtained by computing the missing lid of
this open square:

.. code-block:: text

          x             z
          ^             ^
          |             |
        x |             | q j
          |             |
          x ----------> y
               p i

In the drawing the direction ``i`` goes left-to-right and ``j`` goes

> down-to-up. As we are constructing a path from ``x`` to ``z`` along

Reviewer: ```suggestion
bottom-to-top. As we are constructing a path from ``x`` to ``z`` along
```
Seems slightly better to me like that.

The simplest example of an equivalence is the identity function.

::

  idfun : ∀ {ℓ} → (A : Set ℓ) → A → A
  idfun _ x = x

  idIsEquiv : ∀ {ℓ} (A : Set ℓ) → isEquiv (idfun A)
  equiv-proof (idIsEquiv A) y =
    ((y , refl) , λ z i → z .snd (~ i) , λ j → z .snd (~ i ∨ j))

  idEquiv : ∀ {ℓ} (A : Set ℓ) → A ≃ A
  idEquiv A = (idfun A , idIsEquiv A)

An important special case of equivalent types are isomorphic types
(i.e. types with maps going back and forth which are mutually
inverse): https://github.com/agda/cubical/blob/master/Cubical/Foundations/Isomorphism.agda.

As everything has to work up to higher dimensions the Glue types take
a partial family of types that are equivalent to the base type ``A``:

::

  Glue : ∀ {ℓ ℓ'} (A : Set ℓ) {φ : I}
       → Partial φ (Σ[ T ∈ Set ℓ' ] T ≃ A) → Set ℓ'

..
  ::

  Glue A Te = primGlue A (λ x → Te x .fst) (λ x → Te x .snd)

These come with a constructor and eliminator:

.. code-block:: agda

  glue : ∀ {ℓ ℓ'} {A : Set ℓ} {φ : I} {Te : Partial φ (Σ[ T ∈ Set ℓ' ] T ≃ A)}
       → PartialP φ T → A → Glue A Te

  unglue : ∀ {ℓ ℓ'} {A : Set ℓ} (φ : I) {Te : Partial φ (Σ[ T ∈ Set ℓ' ] T ≃ A)}
         → Glue A Te → A

> 

Reviewer: Maybe write down also the definitional equalities related to `glue` and `unglue` (both when combining them together and the side conditions when `φ` is 1)?

Author: I'm not sure... It is quite technical, but it is also good to know I guess. 

@Saizan What do you think?

Using Glue types we can turn an equivalence of types into a path as
follows:

::

  ua : ∀ {ℓ} {A B : Set ℓ} → A ≃ B → A ≡ B
  ua {_} {A} {B} e i = Glue B (λ { (i = i0) → (A , e)
                                 ; (i = i1) → (B , idEquiv B) })

> The idea is that we glue on ``A`` to ``B`` when ``i = i0`` using ``e``

Reviewer: "we glue on A to B" seems grammatically weird, I had to read it several times to understand what it means. Is "on" needed at all? Maybe it should be "glue A on to B"?

Author: I pushed "The idea is that we glue ``A`` together with ``B`` when ``i = i0`` using ``e`` and ``B`` with itself when ``i = i1`` using the identity equivalence."

Is it clearer?

in order to compute with the equivalences.

For more results about Glue types and univalence see
https://github.com/agda/cubical/blob/master/Cubical/Core/Glue.agda and
https://github.com/agda/cubical/blob/master/Cubical/Foundations/Univalence.agda. For
some examples of what can be done with this for working with binary
and unary numbers see
https://github.com/agda/cubical/blob/master/Cubical/Data/BinNat/BinNat.agda.

Higher inductive types
----------------------

Cubical Agda also lets us directly define higher inductive types as
datatypes with path constructors. For example the circle and `torus
<https://en.wikipedia.org/wiki/Torus>`_ can be defined as:

::

  data S¹ : Set where
    base : S¹
    loop : base ≡ base

  data Torus : Set where
    point : Torus
    line1 : point ≡ point
    line2 : point ≡ point
    square : PathP (λ i → line1 i ≡ line1 i) line2 line2

Functions out of higher inductive types can then be defined using
pattern-matching:

::

  t2c : Torus → S¹ × S¹
  t2c point        = ( base , base )
  t2c (line1 i)    = ( loop i , base )
  t2c (line2 j)    = ( base , loop j )
  t2c (square i j) = ( loop i , loop j )

> 

Reviewer: On the right-hand-sides, the spaces after the opening parentheses and before the closing ones look a bit weird, is there a reason to have them?

Author: Fixed, now the result of t2c looks exactly like the argument to c2t

                            -- (Voevodsky's version of the notion).
           ; _≃_            -- The type of equivalences between two given types.
           ; EquivContr     -- A formulation of univalence.

           ; ∥_∥             -- Propositional truncation.
           ; ∣_∣             -- Map into the propositional truncation.
           ; ∥∥-isProp       -- A truncated type is a proposition.
           ; ∥∥-recursion    -- Non-dependent elimination.
           ; ∥∥-induction    -- Dependent elimination.
           )

In order to get access to only the HoTT/UF primitives start a file as
follows:

.. code-block:: agda

  {-# OPTIONS --cubical #-}

  open import Cubical.Core.HoTT-UF

However, even though this interface exists it is still recommended
that one uses the cubical identity types unless one really need ``J``
to compute on ``refl``. The reason for this is that the syntax for
path types does not work for the identity types, making many proofs
more involved as the only way to reason about them is using ``J``
(furthermore, the path types satisfy many useful definitional
equalities that the identity types don't).

> 

Reviewer: The parenthetical remark could be made into a sentence of its own. It’s actually a very valid reason for avoid identity types, maybe more so than the syntax.

Author: Good point! Done

the general schema for higher inductive types work, following the
`CHM`_ paper.

To use the cubical mode Agda needs to be run with the ``--cubical``
command-line-option or with ``{-# OPTIONS --cubical #-}`` at the top
of the file.

The cubical mode adds the following features to Agda:

1. An interval type and path types
2. Generalized transport (``transp``)
3. Partial elements and systems
4. Homogeneous composition (``hcomp``)
5. Glue types
6. Higher inductive types
7. Cubical identity types

There is a standard ``agda/cubical`` library for Cubical Agda
available at https://github.com/agda/cubical. This documentation uses
the naming conventions of this library, for a detailed list of all of
the built-in Cubical Agda files and primitives see
:ref:`primitives-ref`. The main design choices of the core part of the
library are explained in
https://homotopytypetheory.org/2018/12/06/cubical-agda/.

The recommended way to get access to the Cubical primitives is to add
the following to the top of a file (this assumes that the
``agda/cubical`` library is installed and visible to Agda).

.. code-block:: agda

  {-# OPTIONS --cubical #-}
  module MyModule where

> 

Reviewer: It’s actually optional to start files with `module MyModule where`, so rather than inventing a dummy name, you could just leave out that line.

Author: Oh, I didn't know. I'll remove it

  open import Cubical.Core.Everything

For detailed install instructions for ``agda/cubical`` see:
https://github.com/agda/cubical/blob/master/INSTALL.md. In order to
make this library visible to Agda add
``/path/to/cubical/cubical.agda-lib`` to ``.agda/libraries`` and
``cubical`` to ``.agda/defaults`` (where ``path/to`` is the absolute
path to where the ``agda/cubical`` library has been installed). For
details of Agda's library management see :ref:`package-system`.

> Expert users who doesn't want to rely on ``agda/cubical`` can just add

Reviewer: ```suggestion
Expert users who do not want to rely on ``agda/cubical`` can just add
```

the relevant import statements at the top of their file (for details

> for experts see :ref:`primitives-ref`). However, for beginners it is

Reviewer: ```suggestion
see :ref:`primitives-ref`). However, for beginners it is
```
No need to duplicate `experts`.

https://github.com/Saizan/cubical-demo/. However this is relying on
deprecated features and is not recommended to use.

The interval and path types
---------------------------

The key idea of Cubical Type Theory is to add an interval type ``I :
Setω`` (the reason this is in ``Setω`` is because it doesn't support
the ``transp`` and ``hcomp`` operations). A variable ``i : I``
intuitively corresponds to a point the `real unit interval
<https://en.wikipedia.org/wiki/Unit_interval>`_. In an empty context,
there are only two values of type ``I``: the two endpoints of the
interval, ``i0`` and ``i1``.

.. code-block:: agda

  i0 : I
  i1 : I

Elements of the interval form a `De Morgan algebra
<https://en.wikipedia.org/wiki/De_Morgan_algebra>`_, with minimum
(``∧``), maximum (``∨``) and negation (``~``).

.. code-block:: agda

  _∧_ : I → I → I
  _∨_ : I → I → I
  ~_ : I → I

All the properties of De Morgan algebras hold definitionally. The
endpoints of the interval ``i0`` and ``i1`` are the bottom and top
elements, respectively.

.. code-block:: agda

  module interval-equations (i j : I) where
    data _≡_ (i : I) : I → Setω where
      reflI : i ≡ i

    infix 10 _≡_

> 

Reviewer: Isn’t it confusing to use `_≡_` for the inductive identity type, the same symbol being used for path types a few paragraphs below? I think you could simply list all the equations, with normal equal signs and without giving them a name, and simply say that `=` means definitional equality here. It doesn’t have to be valid Agda code.

Author: Yeah, I'll fix this

Foundations <https://homotopytypetheory.org/>`_. The version of
Cubical Type Theory that Agda implements is a variation of the `CCHM`_
Cubical Type Theory where the Kan composition operations are
decomposed into homogeneous composition and generalized
transport. This is what makes the general schema for higher inductive
types work, following the `CHM`_ paper.

To use the cubical mode Agda needs to be run with the ``--cubical``
command-line-option or with ``{-# OPTIONS --cubical #-}`` at the top
of the file.

The cubical mode adds the following features to Agda:

1. An interval ``I`` type and path types
2. Partial elements and systems
3. Kan operations (``transp`` and ``hcomp``)
4. ``Glue`` types
5. Higher inductive types
6. Cubical identity types

There is a standard library for Cubical Agda available at
https://github.com/agda/cubical.

The main design choices of the core part of the library are explained
in https://homotopytypetheory.org/2018/12/06/cubical-agda/.

In order to get access to the Cubical Agda primitives one should
either import
https://github.com/agda/cubical/blob/master/Cubical/Core/Primitives.agda
or add the relevant import statements from the top of that file.

> 

Reviewer: It would be good to have a code block indicating exactly how to start a file (the `{-# OPTIONS --cubical #-}` and the imports to make). It should be clearly visible as soon as the user opens the doc on cubical Agda, so that people immediately see what they need in order to start experimenting.

Maybe there should be several different ones depending on whether you are using [agda/cubical](https://github.com/agda/cubical), the HoTT-UF version of it, or simply the primitives included in Agda (maybe with suggested renamings), but then it should be made very clear which is which.

Author: I added the recommended one in the beginning now (using agda/cubical) and some pointer for experts. I also added instructions for how to use the HoTT-UF one in the section

There is also an older version of the library available at
https://github.com/Saizan/cubical-demo/.

The interval and path types
---------------------------

The key idea of Cubical Type Theory is to add an interval type ``I :
Setω`` (the reason this is in ``Setω`` is because it doesn't support
the Kan operations). A variable ``i : I`` intuitively corresponds to a
point the `real unit interval
<https://en.wikipedia.org/wiki/Unit_interval>`_. In an empty context,
there are only two values of type ``I``: the two endpoints of the
interval, ``i0`` and ``i1``.

.. code-block:: agda

  i0 : I
  i1 : I

Elements of the interval form a `De Morgan algebra
<https://en.wikipedia.org/wiki/De_Morgan_algebra>`_, with minimum
(``∧``), maximum (``∨``) and negation (``~``).

> 

Reviewer: It should be made clear that various examples use notations from [agda/cubical](https://github.com/agda/cubical)  (for instance `\/` and `/\`), so users know that they may have to download and setup an additional library to try them. For the people who would not want to use [agda/cubical](https://github.com/agda/cubical), the names of all the primitives should be mentioned somehow (maybe in parenthetical remarks when they are introduced, or in a separate section.

Author: I added an appendix listing all of the primitives and built-ins.

    p₇ = reflI
    p₈ = reflI

The core idea of Homotopy Type Theory and Univalent Foundations is a
correspondence between paths (as in topology) and (proof-relevant)
equality (as in Martin-Löf's identity type). This correspondence is
taken very literally in Cubical Agda where a path in a type ``A`` is
represented like a function out of the interval, ``I → A``. In fact a
path type is in fact a special case of the more general built-in
heterogeneous path types:

.. code-block:: agda

   PathP : ∀ {ℓ} (A : I → Set ℓ) → A i0 → A i1 → Set ℓ

   -- Non dependent path types
   Path : ∀ {ℓ} (A : Set ℓ) → A → A → Set ℓ
   Path A a b = PathP (λ _ → A) a b

The central notion of equality in Cubical Agda is hence heterogeneous
(which is sometimes called "John Major equality"). To define paths we

> use λ-abstractions and to apply them we use regular application.  For

Reviewer 2: AFAIU John Major corresponds strictly to Heterogeneous equality + K. The pun being that
we pretend everyone is equal no matter the type but in the end only elements of the same
type are actually treated equally (being substituted). Same as the "classless society" which,
surprise, surprise, happens to still have classes and inequalities.

I think cubical is a different approach: you *expect* a proof that the two endpoints' types are
equal before even talking about them being equal.

Anyway, I think John Major will only add confusion in this context.

Reviewer 3: More than UIP, the issue is where the equality between types resides. John Major is more like this: 
```agda
JM : {A B : Set} -> A -> B -> Set _
JM {A} {B} x y = Σ (p : A ≡ B) \ p → PathP (\ i → p i) x y
```
Which indeed is hard to use for anything without K, because that `p` could be any equivalence.

We could reference the `PathOver` notion instead?

Author: Thanks for clarifying what JM equality is. I was indeed mixing it up with `PathOver`.

  symInv p = refl

  congId : ∀ {ℓ} {A : Set ℓ} {x y : A} (p : x ≡ y) → cong (λ a → a) p ≡ p
  congId p = refl

  congComp : ∀ {ℓ} {A B C : Set ℓ} (f : A → B) (g : B → C) {x y : A} (p : x ≡ y) →
               cong (λ a → g (f a)) p ≡ cong g (cong f p)
  congComp f g p = refl

Path types also lets us prove new things are not provable in standard
Agda, for example function extensionality (pointwise equal functions
are equal) has an extremely simple proof:

.. code-block:: agda

  funExt : ∀ {ℓ} {A : Set ℓ} {B : A → Set ℓ} {f g : (x : A) → B x} →
             ((x : A) → f x ≡ g x) → f ≡ g
  funExt p i x = p x i

Transport
---------

While path types are great for reasoning about equality they don't let
us transport along paths between types or even compose path, which in
particular means that we cannot yet prove the induction principle for
paths. In order to remedy this we also have a built-in (generalized)
transport operation and homogeneous composition operations. The
transport operation is generalized in the sense that it lets us
specify where it is the identity function

.. code-block:: agda

  transp : ∀ {ℓ} (A : I → Set ℓ) (r : I) (a : A i0) → A i1

> When calling ``transp A r a`` Agda makes sure that ``A`` is constant

Reviewer 4: Could someone be so nice as to explain in more specific what "A is constant on r" means? Also, could we have a more intuitive explanation of r?

Author: Hmm, yeah, this is quite subtle and technical. The constraint says that the family `A : I -> Set` is the constant function when r is `i1`. 

For example when typechecking:

```
coe0→i : ∀ {ℓ} (A : I → Set ℓ) (i : I) → A i0 → A i
coe0→i A i a = transp (λ j → A (i ∧ j)) (~ i) a
```

Agda checks that when `(~ i) = i1` then the family is constant. So when `i = i0` we get

```
(λ j → A (i0 ∧ j)) = (λ j → A i0)
```

In particular the type does **not** depend on `j`.

The reason that this is necessary is that when r is i1 the transp function will compute as the identity function:

```
transp (\ j -> A) i1 a = a
```

This of course only works if `A = A (0/j) = A (1/j)`.

Does this help? It is all quite technical and I'm not sure if it should go in the manual... So some help reformulating it in a less technical way would be great.

Author: @riaqn : Did this help? Should I include any of it in the documentation?

Reviewer: I agree that this should be reformulated. Here is a suggestion:

> There is an additional side condition to be satisfied for `transp A r a` to type-check, which is that `A` has to be *constant on `r`*. This means that `A` should be a constant function whenever the constraint `r = i1` is satisfied.
> This side condition is vacuously true when `r` is equal to `i0`, so there is nothing to check when writing `transp A i0 a`. However when `r` is equal to `i1` the `transp` function will compute as the identity function
> ```agda
> transp A i1 a = a
> ```
> and this requires `A` to be constant for it to be well-typed.
> 
> We can use `transp` to define regular transport: […]

Author: Excellent @guillaumebrunerie ! That's much clearer. I'll incorporate it

type, so for example for Sigma types it is computed elementwise. For
path types it is however not yet possible to provide the computation
rule as we need some way to remember the endpoints of the path after
transporting it. Furthermore, this must work for arbitrary higher
dimensional cubes (as we can iterate the path types). For this we
introduce the "homogeneous composition operations" (``hcomp``) that
generalize binary composition of paths to n-ary composition of higher
dimensional cubes.

Partial elements and systems
----------------------------

In order to describe the homogeneous composition operations we need to
be able to write partially specified n-dimensional cubes (i.e. cubes
where some faces are missing). Given an element of the interval ``r :
I`` there is a predicate ``IsOne`` which represents the constraint ``r
= i1``. This comes with a proof that ``i1`` is in fact equal to ``i1``
called ``1=1 : IsOne i1``. We use Greek letters like ``φ`` or ``ψ``
when such an ``r`` should be thought of as being in the image of
``IsOne``.

Using this we introduce a type of partial elements called ``Partial φ
A``, this is a special version of ``IsOne φ → A`` with a more

> extensional judgmental equality. The idea is that ``Partial φ A`` is

Author: @Saizan I think I got the remark about "a more extensional equality" from you. Can you please remind me of what it meant more exactly?

Reviewer 3: Much like in cubicaltt for systems, two elements ``u, t : Partial φ A`` are equal if for every face map `α` that `φ` decomposes into, we have `uα = tα`.

In the docs one might want to be less technical though.

Author: Thanks, I clarified this by saying that u and t are considered equal if they specify the same subcube (so faces can be given in different order, etc.)

.. code-block:: agda

  _[_↦_] : ∀ {ℓ} (A : Set ℓ) (φ : I) (u : Partial φ A) → Setω
  A [ φ ↦ u ] = Sub A φ u

A term ``v : A [ φ ↦ u ]`` is of type ``v : A`` and when ``IsOne φ``
it must be definitionally equal to ``u : A``. Any term ``u : A`` can
be seen as an term of ``A [ φ ↦ u ]`` which agrees with itself on
``φ``:

.. code-block:: agda

  inc : ∀ {ℓ} {A : Set ℓ} {φ : I} (u : A) → A [ φ ↦ (λ _ → u) ]

One can also forget that a partial element agrees with ``u`` on ``φ``:

.. code-block:: agda

  ouc : ∀ {ℓ} {A : Set ℓ} {φ : I} {u : Partial φ A} → A [ φ ↦ u ] → A

With all of this cubical infrastructure we can now describe the
``hcomp`` operations.

Homogeneous composition
-----------------------

The homogeneous composition operations generalize binary composition
of paths so that we can compose multiple composable cubes.

.. code-block:: agda

  hcomp : ∀ {ℓ} {A : Set ℓ} {φ : I} (u : I → Partial φ A) (u0 : A) → A

When calling ``hcomp {φ = φ} u u0`` Agda makes sure that ``u0`` agrees
with ``u i0`` on ``φ``. The idea is that ``u0`` is the base and ``u``
specify the sides of an open box. This is hence an open (higher

> dimensional) cube (maybe with some sides missing) where the side

Reviewer 2: specify -> specifies

Author: Fixed, thanks!

to do a lot of equality reasoning in a very direct way:

..
  ::

  sym : ∀ {ℓ} {A : Set ℓ} {x y : A} → x ≡ y → y ≡ x
  sym p = λ i → p (~ i)

  cong : ∀ {ℓ} {A : Set ℓ} {x y : A} {B : A → Set ℓ} (f : (a : A) → B a) (p : x ≡ y)
         → PathP (λ i → B (p i)) (f x) (f y)
  cong f p i = f (p i)

Because of the way functions compute these satisfy some new
definitional equalities compared to the standard Agda definitions:

..
  ::

  symInv : ∀ {ℓ} {A : Set ℓ} {x y : A} (p : x ≡ y) → sym (sym p) ≡ p
  symInv p = refl

  congId : ∀ {ℓ} {A : Set ℓ} {x y : A} (p : x ≡ y) → cong (λ a → a) p ≡ p
  congId p = refl

  congComp : ∀ {ℓ} {A B C : Set ℓ} (f : A → B) (g : B → C) {x y : A} (p : x ≡ y) →
               cong (λ a → g (f a)) p ≡ cong g (cong f p)
  congComp f g p = refl

Path types also lets us prove new things are not provable in standard
Agda, for example function extensionality (pointwise equal functions
are equal) has an extremely simple proof:

..
  ::

  funExt : ∀ {ℓ} {A : Set ℓ} {B : A → Set ℓ} {f g : (x : A) → B x} →
             ((x : A) → f x ≡ g x) → f ≡ g
  funExt p i x = p x i

Transport
---------

While path types are great for reasoning about equality they don't let

> us transport along paths between types or even compose path, which in

Reviewer: ```suggestion
us transport along paths between types or even compose paths, which in
```

Author: Thanks!

    p₇ = reflI
    p₈ = reflI

The core idea of Homotopy Type Theory and Univalent Foundations is a
correspondence between paths (as in topology) and (proof-relevant)
equality (as in Martin-Löf's identity type). This correspondence is
taken very literally in Cubical Agda where a path in a type ``A`` is

> represented like a function out of the interval, ``I → A``. In fact a

Reviewer: ```suggestion
represented like a function out of the interval, ``I → A``. A
```
duplicate "in fact"

I`` there is a predicate ``IsOne`` which represents the constraint ``r
= i1``. This comes with a proof that ``i1`` is in fact equal to ``i1``
called ``1=1 : IsOne i1``. We use Greek letters like ``φ`` or ``ψ``
when such an ``r`` should be thought of as being in the image of
``IsOne``.

Using this we introduce a type of partial elements called ``Partial φ
A``, this is a special version of ``IsOne φ → A`` with a more
extensional judgmental equality (two elements of ``Partial φ A`` are
considered equal if they represent the same subcube, so the faces of
the cubes can for example be given in different order and the two
elements will still be considered the same). The idea is that
``Partial φ A`` is the type of cubes in ``A`` that are only defined
when ``IsOne φ``.  There is also a dependent version of this called
``PartialP φ A`` which allows ``A`` to be defined only when ``IsOne
φ``.

.. code-block:: agda

  Partial : ∀ {ℓ} → I → Set ℓ → Setω

  PartialP : ∀ {ℓ} → (φ : I) → Partial φ (Set ℓ) → Setω

Partial elements are introduced using pattern matching:

::

  sys : ∀ i → Partial (i ∨ ~ i) Set₁
  sys i (i = i0) = Set
  sys i (i = i1) = Set → Set

> 

Reviewer: You could use `A` instead of `Set₁` and `a_0` and `a_1` instead of `Set` and `Set → Set`, rather than these seemingly random terms. Or even just question marks instead of `Set` and `Set → Set`.

Author: Or maybe even better a concrete type like `Bool`

> The term ``sys`` is hence only defined when ``IsOne (i ∨ ~ i)``, that

Reviewer: Well, no, technically the "term" `sys` is defined no matter what, it’s just that if you want to apply it to some arguments you need to give a term of type `IsOne (i \/ ~ i)`.

Author: Hmm yeah, but the semantic intuition is that `sys` is only defined at the endpoints of the interval.

Author: Btw, I'm removing all mentions of "systems" in favour of the terminology "partial elements" everywhere

Author: I pushed a change now

  idfun _ x = x

  idIsEquiv : ∀ {ℓ} (A : Set ℓ) → isEquiv (idfun A)
  equiv-proof (idIsEquiv A) y =
    ((y , refl) , λ z i → z .snd (~ i) , λ j → z .snd (~ i ∨ j))

  idEquiv : ∀ {ℓ} (A : Set ℓ) → A ≃ A
  idEquiv A = (idfun A , idIsEquiv A)

An important special case of equivalent types are isomorphic types
(i.e. types with maps going back and forth which are mutually
inverse): https://github.com/agda/cubical/blob/master/Cubical/Foundations/Isomorphism.agda.

As everything has to work up to higher dimensions the Glue types take
a partial family of types that are equivalent to the base type ``A``:

.. code-block:: agda

  primGlue : ∀ {ℓ ℓ'} (A : Set ℓ) {φ : I} (T : Partial φ (Set ℓ'))
           → (e : PartialP φ (λ o → T o ≃ A)) → Set ℓ'

These come with a constructor and eliminator:

.. code-block:: agda

  glue : ∀ {ℓ ℓ'} {A : Set ℓ} {φ : I} {T : Partial φ (Set ℓ')}
           {e : PartialP φ (λ o → T o ≃ A)}
         → PartialP φ T → A → primGlue A T e

  unglue : ∀ {ℓ ℓ'} {A : Set ℓ} (φ : I) {T : Partial φ (Set ℓ')}
             {e : PartialP φ (λ o → T o ≃ A)} → primGlue A T e → A

In the cubical library we uncurry the Glue types to make them more
pleasant to use:

> 

Reviewer: Given that you expect the users to use agda/cubical, it is rather irrelevant here that `Glue` and `primGlue` are different, you should talk only about `Glue` and mention the difference in currification in the appendix.

Author: Good point! I didn't like it when I wrote this...
